metaflow/plugins/docker/docker_environment.py

```diff
@@ -0,0 +1,135 @@
+import os
+
+from metaflow.exception import MetaflowException
+from metaflow.metaflow_config import (
+    _USE_BAKERY,
+)
+from metaflow.metaflow_environment import MetaflowEnvironment
+from metaflow.plugins.pypi.conda_environment import CondaEnvironment
+from .fast_bakery import bake_image
+from metaflow.plugins.aws.batch.batch_decorator import BatchDecorator
+from metaflow.plugins.kubernetes.kubernetes_decorator import KubernetesDecorator
+from metaflow.plugins.pypi.conda_decorator import CondaStepDecorator
+
+
+class DockerEnvironmentException(MetaflowException):
+    headline = "Ran into an error while setting up the environment"
+
+    def __init__(self, msg):
+        super(DockerEnvironmentException, self).__init__(msg)
+
+
+class DockerEnvironment(MetaflowEnvironment):
+    TYPE = "docker"
+    _filecache = None
+
+    def __init__(self, flow):
+        self.flow = flow
+
+    def set_local_root(self, local_root):
+        # TODO: Make life simple by passing echo to the constructor and getting rid of
+        # this method's invocation in the decorator
+        self.local_root = local_root
+
+    def decospecs(self):
+        # Apply conda decorator to manage the task execution lifecycle.
+        return ("conda",) + super().decospecs()
+
+    def validate_environment(self, echo, datastore_type):
+        self.datastore_type = datastore_type
+        self.echo = echo
+
+        # Avoiding circular imports.
+        from metaflow.plugins import DATASTORES
+
+        self.datastore = [d for d in DATASTORES if d.TYPE == self.datastore_type][0]
+
+        # Use remote image bakery for conda environments if configured.
+        if not _USE_BAKERY:
+            raise DockerEnvironmentException("Image Bakery is not configured.")
+
+    def _setup_conda_fallback(self):
+        # TODO: In the future we want to support executing with Docker even locally.
+        have_to_delegate = False
+        for step in self.flow:
+            if not any(_is_remote_deco(deco) for deco in step.decorators):
+                # We need to fall back to the Conda environmment for flows that are trying to execute anything locally.
+                have_to_delegate = True
+        if not have_to_delegate:
+            return False
+        # replace methods with the delegate class that we determined.
+        cls = CondaEnvironment
+        for k in cls.__dict__:
+            obj = getattr(cls, k)
+            if callable(obj):
+                setattr(self.__class__, k, obj)
+        return True
+
+    def init_environment(self, echo):
+        if self._setup_conda_fallback():
```

Comment:
it seems that we fall back to conda environments for all steps if any step is running locally. can we make it such that any step that is marked to run on kubernetes can benefit from fast bakery?